lib/microsoft/office.rb

```diff
@@ -0,0 +1,1153 @@
+            attendees.push({
+                type: "resource",
+                emailAddress: {
+                    address: room.email,
+                    name: room.name
+                }
+            })
+        end
+
+        # Add the current user as an attendee
+        if current_user && add_current_user
+            attendees.push({
+                emailAddress: {
+                    address: current_user[:email],
+                    name: current_user[:name]
+                }
+            })
+        end
+
+        # Create our event which will eventually be stringified
+        event = {
+            subject: subject,
+            body: {
+                contentType: content_type,
+                content: "X!X!X!"
+            },
+            start: {
+                dateTime: start_param,
+                timeZone: timezone
+            },
+            end: {
+                dateTime: end_param,
+                timeZone: timezone
+            },
+            location: {
+                displayName: rooms.map{|room| room.name}.join(" and "),
+                locationEmailAddress: rooms[0].id
+            },
+            isOrganizer: false,
+            attendees: attendees
+        }
+        event[:location][:displayName] = location if location
+        exts = []
+        extensions.each do |extension|
+            ext = {
+                "@odata.type": "microsoft.graph.openTypeExtension",
+                "extensionName": extension[:name]
+            }
+            extension[:values].each do |ext_key, ext_value|
+                ext[ext_key] = ext_value
+            end
+            exts.push(ext)
+        end
+        event[:extensions] = exts
+
+        if current_user
+            event[:organizer] = {
+                emailAddress: {
+                    address: current_user[:email],
+                    name: current_user[:name]
+                }
+            }
+        else
+            event[:organizer] = {
+                emailAddress: {
+                    address: rooms[0].email,
+                    name: rooms[0].name
+                }
+            }
+        end
+
+        if recurrence
+            recurrence_range = {
+                type: 'endDate',
+                startDate: start_object.strftime("%F"),
+                endDate: recurrence_end.strftime("%F")
+            }
+            STDERR.puts "Creating recurrence with details:"
+            STDERR.puts recurrence_range
+            STDERR.puts "In timezone: #{timezone}"
+            STDERR.flush
+            event[:recurrence] = {
+                pattern: {
+                    type: recurrence,
+                    interval: 1,
+                    daysOfWeek: [start_object.strftime("%A")]
+                },
+                range: recurrence_range
+            }
+        end
+
+        if is_private
+            event[:sensitivity] = 'private'
+        end
+
+        event = event.to_json
+
+        event.gsub!("X!X!X!",description)
```

Comment:
Can this be inserted directly into attributes that need the substitution before stringifying to JSON? Should cut down the amount of string manipulation needed.